File: artifacts/bst/decompose/src/bst.c
struct node *newNode(int item)
{
  struct node *temp = (struct node *) malloc(sizeof(struct node));
  temp->key = item;
  temp->left = 0;
  temp->right = 0;
  return temp;
}


/*** DEPENDENCIES:
struct node
{
  int key;
  struct node *left;
  struct node *right;
}
----------------------------
***/


struct node *minValueNode(struct node *node)
{
  struct node *current = node;
  unsigned int current_idx = 0;
  while ((&current[current_idx]) && (current->left != 0))
    current_idx = current_idx->left;

  return current;
}


/*** DEPENDENCIES:
struct node
{
  int key;
  struct node *left;
  struct node *right;
}
----------------------------
***/


void inorder(struct node *root)
{
  if (root != 0)
  {
    inorder(root->left);
    printf("%d ", root->key);
    inorder(root->right);
  }
}


/*** DEPENDENCIES:
struct node
{
  int key;
  struct node *left;
  struct node *right;
}
----------------------------
***/


struct node *insert(struct node *node, int key)
{
  if (node == 0)
  {
    return newNode(key);
  }
  if (key < node->key)
  {
    node->left = insert(node->left, key);
  }
  else
    node->right = insert(node->right, key);
  return node;
}


/*** DEPENDENCIES:
struct node
{
  int key;
  struct node *left;
  struct node *right;
}
----------------------------
struct node *newNode(int item)
{
  struct node *temp = (struct node *) malloc(sizeof(struct node));
  temp->key = item;
  temp->left = 0;
  temp->right = 0;
  return temp;
}


----------------------------
***/


struct node *deleteNode(struct node *root, int key)
{
  if (root == 0)
  {
    return root;
  }
  if (key < root->key)
  {
    root->left = deleteNode(root->left, key);
  }
  else
    if (key > root->key)
  {
    root->right = deleteNode(root->right, key);
  }
  else
  {
    if (root->left == 0)
    {
      struct node *temp = root->right;
      free(root);
      return temp;
    }
    else
      if (root->right == 0)
    {
      struct node *temp = root->left;
      free(root);
      return temp;
    }
    struct node *temp = minValueNode(root->right);
    root->key = temp->key;
    root->right = deleteNode(root->right, temp->key);
  }
  return root;
}


/*** DEPENDENCIES:
struct node
{
  int key;
  struct node *left;
  struct node *right;
}
----------------------------
struct node *minValueNode(struct node *node)
{
  struct node *current = node;
  unsigned int current_idx = 0;
  while ((&current[current_idx]) && (current->left != 0))
    current_idx = current_idx->left;

  return current;
}


----------------------------
***/

File: artifacts/bst/translate/src/bst.rs
use crate::*;
use std::mem;
use std::fmt::Write;

pub fn newNode(item: i32) -> Option<Box<Node>> {
    Some(Box::new(Node {
        key: item,
        left: None,
        right: None,
    }))
}
pub fn min_value_node(node: Option<Box<Node>>) -> Option<Box<Node>> {
    let mut current = node;
    while let Some(ref mut current_node) = current {
        if current_node.left.is_some() {
            current = current_node.left.take();
        } else {
            break;
        }
    }
    current
}

pub fn inorder(root: Option<&Box<Node>>) {
    if let Some(node) = root {
        inorder(node.left.as_ref());
        print!("{} ", node.key);
        inorder(node.right.as_ref());
    }
}
pub fn insert(node: Option<Box<Node>>, key: i32) -> Option<Box<Node>> {
    match node {
        None => newNode(key),
        Some(mut node) => {
            if key < node.key {
                node.left = insert(node.left, key);
            } else {
                node.right = insert(node.right, key);
            }
            Some(node)
        }
    }
}

pub fn delete_node(root: Option<Box<Node>>, key: i32) -> Option<Box<Node>> {
    let mut root = match root {
        Some(node) => node,
        None => return None,
    };

    if key < root.key {
        root.left = delete_node(root.left, key);
    } else if key > root.key {
        root.right = delete_node(root.right, key);
    } else {
        if root.left.is_none() {
            return root.right;
        } else if root.right.is_none() {
            return root.left;
        }

        let min_right = min_value_node(root.right.clone()).unwrap();
        root.key = min_right.key;
        root.right = delete_node(root.right, min_right.key);
    }

    Some(root)
}
